Using Hibernate 5.x and Oracle driver throws IllegalAccessException.

Using Hibernate 5.x and Oracle driver throws IllegalAccessException.

## Environment Details
* GlassFish Version: 5.0-b08
* JDK version: 1.8.0_65
* OS: Windows 7
* Database: Oracle 12c Release 1

----------

## Problem Description
Using Hibernate 5.x and Oracle driver throws java.lang.IllegalAccessException below.
```
[2017-06-22T10:45:05.148+0900] [glassfish 5.0] [SEVERE] [jdbc.ex_connection_holder] [javax.enterprise.resource.resourceadapter.com.sun.gjc.spi] [tid: _ThreadID=45 _ThreadName=admin-listener(3)] [timeMillis: 1498095905148] [levelValue: 1000] [[
  RAR9008 : Exception while invoking method on Connection object. Exception : {0}
javax.resource.ResourceException: java.lang.IllegalAccessException: Class com.sun.gjc.util.MethodExecutor can not access a member of class oracle.jdbc.driver.PhysicalConnection with modifiers "public"
	at com.sun.gjc.util.MethodExecutor.invokeMethod(MethodExecutor.java:240)
	at com.sun.gjc.spi.jdbc40.ConnectionHolder40.getSchema(ConnectionHolder40.java:609)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.hibernate.engine.jdbc.env.internal.DefaultSchemaNameResolver$SchemaNameResolverJava17Delegate.resolveSchemaName(DefaultSchemaNameResolver.java:72)
	at org.hibernate.engine.jdbc.env.internal.DefaultSchemaNameResolver.resolveSchemaName(DefaultSchemaNameResolver.java:59)
	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentImpl.determineCurrentSchemaName(JdbcEnvironmentImpl.java:298)
	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentImpl.<init>(JdbcEnvironmentImpl.java:232)
	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:114)
	at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator.initiateService(JdbcEnvironmentInitiator.java:35)
	at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:88)
	at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:259)
	at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:233)
	at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:210)
	at org.hibernate.engine.jdbc.internal.JdbcServicesImpl.configure(JdbcServicesImpl.java:51)
	at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.configureService(StandardServiceRegistryImpl.java:94)
	at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:242)
	at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:210)
	at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.handleTypes(MetadataBuildingProcess.java:352)
	at org.hibernate.boot.model.process.spi.MetadataBuildingProcess.complete(MetadataBuildingProcess.java:111)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.metadata(EntityManagerFactoryBuilderImpl.java:858)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:885)
	at org.hibernate.jpa.HibernatePersistenceProvider.createContainerEntityManagerFactory(HibernatePersistenceProvider.java:151)
	at org.glassfish.persistence.jpa.PersistenceUnitLoader.loadPU(PersistenceUnitLoader.java:199)
	at org.glassfish.persistence.jpa.PersistenceUnitLoader.<init>(PersistenceUnitLoader.java:107)
	at org.glassfish.persistence.jpa.JPADeployer$1.visitPUD(JPADeployer.java:223)
	at org.glassfish.persistence.jpa.JPADeployer$PersistenceUnitDescriptorIterator.iteratePUDs(JPADeployer.java:510)
	at org.glassfish.persistence.jpa.JPADeployer.createEMFs(JPADeployer.java:230)
	at org.glassfish.persistence.jpa.JPADeployer.prepare(JPADeployer.java:168)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.prepareModule(ApplicationLifecycle.java:925)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:434)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:219)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:491)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:540)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:536)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:535)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:566)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:558)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:557)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1465)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:110)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1847)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1723)
	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:408)
	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:147)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:183)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:192)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:102)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:491)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:416)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:105)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:281)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:377)
	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:174)
	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:179)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:460)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:168)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:242)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalAccessException: Class com.sun.gjc.util.MethodExecutor can not access a member of class oracle.jdbc.driver.PhysicalConnection with modifiers "public"
	at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102)
	at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(AccessibleObject.java:296)
	at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:288)
	at java.lang.reflect.Method.invoke(Method.java:490)
	at com.sun.gjc.util.MethodExecutor.invokeMethod(MethodExecutor.java:238)
	... 93 more
]]
```

In MethodExecutor#invokeMethod(Object object, String methodName, Class<?>[] valueTypes, Object... values) method, object.getClass().getMethod(methodName, valueTypes) returns the Method object of the 
jdbc driver internal class.
If jdbc driver internal class is not public, its Method object can't be invoked as it is not accessible.

If using Hibernate 5.x and Oracle driver, ConnectionHolder40#getSchema() method calls MethodExecutor#invokeMethod with Oracle internal T4CConnection class as object argument.
In MethodExecutor#invokeMethod, T4CConnection.getClass().getMethod("getSchema") returns a getSchema method of Oracle internal PhysicalConnection class.
Because PhysicalConnection class is not public class, invoking its getSchema method throws IllegalAccessException.

In payara, same problem issued at https://github.com/payara/Payara/issues/730

## Steps to reproduce
1. Download [hibernate5-oracle.zip](https://github.com/javaee/glassfish/files/1093445/hibernate5-oracle.zip)
1. Unzip and cd hibernate5-oracle
1. Run `mvn install` to build hibernate5-oracle-1.0-SNAPSHOT.war
1. Download Oracle JDBC Driver (ojdbc7.jar) from http://www.oracle.com/technetwork/database/features/jdbc/jdbc-drivers-12c-download-1958347.html
1. Copy ojdbc7.jar to glassfish5/glassfish/domains/domain1/lib/
1. Run `asadmin start-domain`
1. Run `asadmin create-jdbc-connection-pool --datasourceclassname oracle.jdbc.pool.OracleDataSource --property user=<user>:password=<password>:url=jdbc\:oracle\:thin\:@//<serverName>\:<portNumber>/<serviceName> oracle`
1. Run `asadmin create-jdbc-resource --connectionpoolid oracle jdbc/user`
1. Run `asadmin deploy /path/to/hibernate5-oracle-1.0-SNAPSHOT.war`

## Fix
There are two ways to fix this issue.
1. Call method.setAccessible(true) before invoking method object.
1. Call getMethod to public interface class instead of jdbc driver internal class.

I think 2. is better because 1. requires security manager policy if security manager exists.

### Call method.setAccessible(true) before invoking method object.
Calling method.setAccessible(true) suppresses Java language access checking when it is used.
So calling method.setAccessible(true) before invoking method object would fix this issue.

```diff
diff --git "a/C:\\Users\\uno\\AppData\\Local\\Temp\\MetA2BA.tmp\\MethodExecutor-HEAD-left.java" "b/D:\\repo\\external\\glassfish\\appserver\\jdbc\\jdbc-ra\\jdbc-core\\src\\main\\java\\com\\sun\\gjc\\util\\MethodExecutor.java"
index 715e1a1..46401da 100644
--- "a/C:\\Users\\uno\\AppData\\Local\\Temp\\MetA2BA.tmp\\MethodExecutor-HEAD-left.java"
+++ "b/D:\\repo\\external\\glassfish\\appserver\\jdbc\\jdbc-ra\\jdbc-core\\src\\main\\java\\com\\sun\\gjc\\util\\MethodExecutor.java"
@@ -92,6 +92,7 @@ public class MethodExecutor implements java.io.Serializable {
             if (parameters.length == 1) {
                 Object[] values = new Object[1];
                 values[0] = convertType(parameters[0], value);
+                method.setAccessible(true);
                 method.invoke(obj, values);
             }
         } catch (IllegalAccessException iae) {
@@ -109,6 +110,11 @@ public class MethodExecutor implements java.io.Serializable {
             _logger.log(Level.SEVERE, "", ite);
             String msg = sm.getString("me.access_denied", method.getName());
             throw new ResourceException(msg);
+        } catch (SecurityException se) {
+            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
+            _logger.log(Level.SEVERE, "", se);
+            String msg = sm.getString("me.access_denied", method.getName());
+            throw new ResourceException(msg);
         }
     }
 
@@ -137,6 +143,7 @@ public class MethodExecutor implements java.io.Serializable {
                     actualValues[i] = convertType(parameters[i], val);
                 }
             }
+            method.setAccessible(true);
             method.invoke(obj, actualValues);
         } catch (IllegalAccessException iae) {
             _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
@@ -153,6 +160,11 @@ public class MethodExecutor implements java.io.Serializable {
             _logger.log(Level.SEVERE, "", ite);
             String msg = sm.getString("me.access_denied", method.getName());
             throw new ResourceException(msg);
+        } catch (SecurityException se) {
+            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
+            _logger.log(Level.SEVERE, "", se);
+            String msg = sm.getString("me.access_denied", method.getName());
+            throw new ResourceException(msg);
         }
     }
 
@@ -235,6 +247,7 @@ public class MethodExecutor implements java.io.Serializable {
         }
         if (actualMethod != null) {
             try {
+                actualMethod.setAccessible(true);
                 returnValue = actualMethod.invoke(object, values);
             } catch (IllegalAccessException ex) {
                 throw new ResourceException(ex);
@@ -242,6 +255,8 @@ public class MethodExecutor implements java.io.Serializable {
                 throw new ResourceException(ex);
             } catch (InvocationTargetException ex) {
                 throw new ResourceException(ex);
+            } catch (SecurityException ex) {
+                throw new ResourceException(ex);
             }
         }
         return returnValue;
```

But if there is a security manager, calling method.setAccessible(true) is checked by security manager that ReflectPermission("suppressAccessChecks") permission is granted.

Glassfish default security policy file (server.policy) grants AllPermission to glassfish internal classes.
```
// classes in lib get all permissions by default
grant codeBase "file:${com.sun.aas.installRoot}/lib/-" {
    permission java.security.AllPermission;
};
```
So this fix works fine unless above policy is deleted.

### Call getMethod to public interface class object instead of jdbc driver internal class.
Calling getMethod to jdbc driver internal class is not appropriate because that class may not be public.
Instead, calling getMethod to public interface class that class implements is appropriate.

```diff
 .../src/main/java/com/sun/gjc/util/MethodExecutor.java         |    4 ++--
 .../main/java/com/sun/gjc/spi/jdbc40/ConnectionHolder40.java   |    8 ++++----
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
index 715e1a1..3c77835 100644
--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
@@ -222,12 +222,12 @@ public class MethodExecutor implements java.io.Serializable {
         }
     }
 
-    public Object invokeMethod(Object object, String methodName, 
+    public Object invokeMethod(Object object, Class<?> interfaceType, String methodName, 
             Class<?>[] valueTypes, Object... values) throws ResourceException {
         Object returnValue = null;
         Method actualMethod = null;
         try {
-            actualMethod = object.getClass().getMethod(methodName, valueTypes);
+            actualMethod = interfaceType.getMethod(methodName, valueTypes);
         } catch (NoSuchMethodException ex) {
             throw new ResourceException(ex);
         } catch (SecurityException ex) {
diff --git a/appserver/jdbc/jdbc-ra/jdbc40/src/main/java/com/sun/gjc/spi/jdbc40/ConnectionHolder40.java b/appserver/jdbc/jdbc-ra/jdbc40/src/main/java/com/sun/gjc/spi/jdbc40/ConnectionHolder40.java
index 5ab5e81..5600dfe 100644
--- a/appserver/jdbc/jdbc-ra/jdbc40/src/main/java/com/sun/gjc/spi/jdbc40/ConnectionHolder40.java
+++ b/appserver/jdbc/jdbc-ra/jdbc40/src/main/java/com/sun/gjc/spi/jdbc40/ConnectionHolder40.java
@@ -592,7 +592,7 @@ public class ConnectionHolder40 extends ConnectionHolder {
             checkValidity();
             Class<?>[] valueTypes = new Class<?>[]{String.class};
             try {
-                getMethodExecutor().invokeMethod(con, "setSchema", valueTypes, schema);
+                getMethodExecutor().invokeMethod(con, Connection.class, "setSchema", valueTypes, schema);
             } catch (ResourceException ex) {
                 _logger.log(Level.SEVERE, "jdbc.ex_connection_holder", ex);
                 throw new SQLException(ex);
@@ -606,7 +606,7 @@ public class ConnectionHolder40 extends ConnectionHolder {
         if(DataSourceObjectBuilder.isJDBC41()) {
             checkValidity();
             try {
-                return (String) getMethodExecutor().invokeMethod(con, "getSchema", null);
+                return (String) getMethodExecutor().invokeMethod(con, Connection.class, "getSchema", null);
             } catch (ResourceException ex) {
                 _logger.log(Level.SEVERE, "jdbc.ex_connection_holder", ex);
                 throw new SQLException(ex);
@@ -621,7 +621,7 @@ public class ConnectionHolder40 extends ConnectionHolder {
             checkValidity();
             Class<?>[] valueTypes = new Class<?>[]{Executor.class, Integer.TYPE};
             try {
-                getMethodExecutor().invokeMethod(con, "setNetworkTimeout", valueTypes, executorObj, milliseconds);
+                getMethodExecutor().invokeMethod(con, Connection.class, "setNetworkTimeout", valueTypes, executorObj, milliseconds);
             } catch (ResourceException ex) {
                 _logger.log(Level.SEVERE, "jdbc.ex_connection_holder", ex);
                 throw new SQLException(ex);
@@ -635,7 +635,7 @@ public class ConnectionHolder40 extends ConnectionHolder {
         if (DataSourceObjectBuilder.isJDBC41()) {
             checkValidity();
             try {
-                return (Integer) getMethodExecutor().invokeMethod(con, "getNetworkTimeout", null);
+                return (Integer) getMethodExecutor().invokeMethod(con, Connection.class, "getNetworkTimeout", null);
             } catch (ResourceException ex) {
                 _logger.log(Level.SEVERE, "jdbc.ex_connection_holder", ex);
                 throw new SQLException(ex);
```
Above is not complete diff.
This fix changes signature of MethodExecutor#invokeMethod method, classes using that method are also changed to adjust new signature.

In my search, these classes uses MethodExecutor#invokeMethod 
* com.sun.gjc.spi.jdbc40.ConnectionHolder40
* com.sun.gjc.spi.jdbc40.DataSource40
* com.sun.gjc.spi.base.StatementWrapper
* com.sun.gjc.spi.jdbc40.CallableStatementWrapper40
* com.sun.gjc.spi.jdbc40.DatabaseMetaDataWrapper40
* com.sun.gjc.spi.jdbc40.ResultSetWrapper40

--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
@@ -92,9 +92,10 @@ public class MethodExecutor implements java.io.Serializable {
             if (parameters.length == 1) {
                 Object[] values = new Object[1];
                 values[0] = convertType(parameters[0], value);
+                method.setAccessible(true);
                 method.invoke(obj, values);
             }
-        } catch (IllegalAccessException iae) {
+        } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
             _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
             _logger.log(Level.SEVERE, "", iae);
             String msg = sm.getString("me.access_denied", method.getName());
@@ -103,11 +104,6 @@ public class MethodExecutor implements java.io.Serializable {
             _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
             _logger.log(Level.SEVERE, "", ie);
             String msg = sm.getString("me.illegal_args", method.getName());
-            throw new ResourceException(msg);
-        } catch (InvocationTargetException ite) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
-            _logger.log(Level.SEVERE, "", ite);
-            String msg = sm.getString("me.access_denied", method.getName());
             throw new ResourceException(msg);
         }
     }
@@ -137,8 +133,9 @@ public class MethodExecutor implements java.io.Serializable {
                     actualValues[i] = convertType(parameters[i], val);
                 }
             }
+            method.setAccessible(true);
             method.invoke(obj, actualValues);
-        } catch (IllegalAccessException iae) {
+        } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
             _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
             _logger.log(Level.SEVERE, "", iae);
             String msg = sm.getString("me.access_denied", method.getName());
@@ -147,11 +144,6 @@ public class MethodExecutor implements java.io.Serializable {
             _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
             _logger.log(Level.SEVERE, "", ie);
             String msg = sm.getString("me.illegal_args", method.getName());
-            throw new ResourceException(msg);
-        } catch (InvocationTargetException ite) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
-            _logger.log(Level.SEVERE, "", ite);
-            String msg = sm.getString("me.access_denied", method.getName());
             throw new ResourceException(msg);
         }
     }
@@ -235,12 +227,11 @@ public class MethodExecutor implements java.io.Serializable {
         }
         if (actualMethod != null) {
             try {
+                actualMethod.setAccessible(true);
                 returnValue = actualMethod.invoke(object, values);
-            } catch (IllegalAccessException ex) {
+            } catch (IllegalAccessException | InvocationTargetException | SecurityException ex) {
                 throw new ResourceException(ex);
             } catch (IllegalArgumentException ex) {
-                throw new ResourceException(ex);
-            } catch (InvocationTargetException ex) {
                 throw new ResourceException(ex);
             }
         }

--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
@@ -263,7 +263,7 @@ public class MethodExecutor implements java.io.Serializable {
                     });
             } catch (PrivilegedActionException e) {
                 if(e.getException() != null){
-                    throw (ResourceException)e.getException();
+                    throw new ResourceException(e.getException());
                 }else{
                     throw new ResourceException(e);
                 }

--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
@@ -49,6 +49,8 @@ import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
 import java.util.Vector;
 import java.util.Properties;
 import java.util.logging.Level;
@@ -253,20 +255,18 @@ public class MethodExecutor implements java.io.Serializable {
             throw new ResourceException(ex);
         }
         if (actualMethod != null) {
-            final ResourceException[] exception = new ResourceException[1];
-            returnValue = AccessController.doPrivileged(new PrivilegedAction() {
-                public Object run() {
-                    try {
-                        actualMethod.setAccessible(true);
-                        return actualMethod.invoke(object, values);
-                    } catch (IllegalAccessException | InvocationTargetException | SecurityException | IllegalArgumentException ex) {
-                        exception[0] =  new ResourceException(ex);
-                    }
-                    return null;
+            try {
+                returnValue = AccessController.doPrivileged(
+                    (PrivilegedExceptionAction<Object>) () -> {
+                            actualMethod.setAccessible(true);
+                            return actualMethod.invoke(object, values);
+                    });
+            } catch (PrivilegedActionException e) {
+                if(e.getException() != null){
+                    throw (ResourceException)e.getException();
+                }else{
+                    throw new ResourceException(e);
                 }
-            });
-            if( exception[0] != null){
-                throw exception[0];
             }
         }
         return returnValue;

--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/util/MethodExecutor.java
@@ -47,6 +47,8 @@ import com.sun.logging.LogDomains;
 import javax.resource.ResourceException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 import java.util.Vector;
 import java.util.Properties;
 import java.util.logging.Level;
@@ -87,24 +89,38 @@ public class MethodExecutor implements java.io.Serializable {
         if (value == null || value.trim().equals("")) {
             return;
         }
-        try {
-            Class[] parameters = method.getParameterTypes();
-            if (parameters.length == 1) {
-                Object[] values = new Object[1];
-                values[0] = convertType(parameters[0], value);
-                method.setAccessible(true);
-                method.invoke(obj, values);
+
+        Class[] parameters = method.getParameterTypes();
+        if (parameters.length == 1) {
+            Object[] values = new Object[1];
+            values[0] = convertType(parameters[0], value);
+
+            final ResourceException[] exception = new ResourceException[1];
+            AccessController
+                .doPrivileged(new PrivilegedAction() {
+                    public Object run() {
+                        try {
+                            method.setAccessible(true);
+                            method.invoke(obj, values);
+                        } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
+                            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
+                            _logger.log(Level.SEVERE, "", iae);
+                            String msg = sm.getString("me.access_denied",
+                                method.getName());
+                            exception[0] = new ResourceException(msg);
+                        } catch (IllegalArgumentException ie) {
+                            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
+                            _logger.log(Level.SEVERE, "", ie);
+                            String msg = sm
+                                .getString("me.illegal_args", method.getName());
+                            exception[0] = new ResourceException(msg);
+                        }
+                        return null;
+                    }
+                });
+            if( exception[0] != null){
+                throw exception[0];
             }
-        } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
-            _logger.log(Level.SEVERE, "", iae);
-            String msg = sm.getString("me.access_denied", method.getName());
-            throw new ResourceException(msg);
-        } catch (IllegalArgumentException ie) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", value);
-            _logger.log(Level.SEVERE, "", ie);
-            String msg = sm.getString("me.illegal_args", method.getName());
-            throw new ResourceException(msg);
         }
     }
 
@@ -119,7 +135,6 @@ public class MethodExecutor implements java.io.Serializable {
      *          a security violation.
      */
     public void runMethod(Method method, Object obj, Vector values) throws ResourceException {
-        try {
             Class[] parameters = method.getParameterTypes();
             if (values.size() != parameters.length) {
                 return;
@@ -133,18 +148,30 @@ public class MethodExecutor implements java.io.Serializable {
                     actualValues[i] = convertType(parameters[i], val);
                 }
             }
-            method.setAccessible(true);
-            method.invoke(obj, actualValues);
-        } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
-            _logger.log(Level.SEVERE, "", iae);
-            String msg = sm.getString("me.access_denied", method.getName());
-            throw new ResourceException(msg);
-        } catch (IllegalArgumentException ie) {
-            _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
-            _logger.log(Level.SEVERE, "", ie);
-            String msg = sm.getString("me.illegal_args", method.getName());
-            throw new ResourceException(msg);
+        final ResourceException[] exception = new ResourceException[1];
+         AccessController.doPrivileged(new PrivilegedAction() {
+            public Object run() {
+                try {
+                    method.setAccessible(true);
+                    method.invoke(obj, actualValues);
+                } catch (IllegalAccessException | InvocationTargetException | SecurityException iae) {
+                    _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
+                    _logger.log(Level.SEVERE, "", iae);
+                    String msg = sm
+                        .getString("me.access_denied", method.getName());
+                    exception[0] = new ResourceException(msg);
+                } catch (IllegalArgumentException ie) {
+                    _logger.log(Level.SEVERE, "jdbc.exc_jb_val", values);
+                    _logger.log(Level.SEVERE, "", ie);
+                    String msg = sm
+                        .getString("me.illegal_args", method.getName());
+                    exception[0] = new ResourceException(msg);
+                }
+                return null;
+            }
+        });
+        if( exception[0] != null){
+            throw exception[0];
         }
     }
 
@@ -217,7 +244,7 @@ public class MethodExecutor implements java.io.Serializable {
     public Object invokeMethod(Object object, String methodName, 
             Class<?>[] valueTypes, Object... values) throws ResourceException {
         Object returnValue = null;
-        Method actualMethod = null;
+        Method actualMethod ;
         try {
             actualMethod = object.getClass().getMethod(methodName, valueTypes);
         } catch (NoSuchMethodException ex) {
@@ -226,13 +253,20 @@ public class MethodExecutor implements java.io.Serializable {
             throw new ResourceException(ex);
         }
         if (actualMethod != null) {
-            try {
-                actualMethod.setAccessible(true);
-                returnValue = actualMethod.invoke(object, values);
-            } catch (IllegalAccessException | InvocationTargetException | SecurityException ex) {
-                throw new ResourceException(ex);
-            } catch (IllegalArgumentException ex) {
-                throw new ResourceException(ex);
+            final ResourceException[] exception = new ResourceException[1];
+            returnValue = AccessController.doPrivileged(new PrivilegedAction() {
+                public Object run() {
+                    try {
+                        actualMethod.setAccessible(true);
+                        return actualMethod.invoke(object, values);
+                    } catch (IllegalAccessException | InvocationTargetException | SecurityException | IllegalArgumentException ex) {
+                        exception[0] =  new ResourceException(ex);
+                    }
+                    return null;
+                }
+            });
+            if( exception[0] != null){
+                throw exception[0];
             }
         }
         return returnValue;

